File: examples/governance/model_registry_domain/lib/model_registry_domain/data_usage_agreement/commands/renew_agreement.rb
module ModelRegistryDomain
  class DataUsageAgreement
    module Commands
      class RenewAgreement
        include Hecks::Command
        emits "RenewedAgreement"

        attr_reader :agreement_id, :expiration_date

        def initialize(agreement_id: nil, expiration_date: nil)
          @agreement_id = agreement_id
          @expiration_date = expiration_date
        end

        def call
          existing = repository.find(agreement_id)
          if existing
            unless ["active", "revoked"].include?(existing.status)
              raise Hecks::Error, "Cannot RenewAgreement: status must be one of active, revoked, got '#{existing.status}'"
            end
            DataUsageAgreement.new(
              id: existing.id,
              model_id: existing.model_id,
              data_source: existing.data_source,
              purpose: existing.purpose,
              consent_type: existing.consent_type,
              effective_date: existing.effective_date,
              expiration_date: expiration_date,
              restrictions: existing.restrictions,
              status: "active"
            )
          else
            raise Hecks::Error, "DataUsageAgreement not found: #{agreement_id}"
          end
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/renew_agreement.go
package domain

import (
	"time"
	"fmt"
)

type RenewAgreement struct {
	AgreementId string `json:"agreement_id"`
	ExpirationDate time.Time `json:"expiration_date"`
}

func (c RenewAgreement) CommandName() string { return "RenewAgreement" }

func (c RenewAgreement) Execute(repo DataUsageAgreementRepository) (*DataUsageAgreement, *RenewedAgreement, error) {
	existing, err := repo.Find(c.AgreementId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("DataUsageAgreement not found: %s", c.AgreementId)
	}
	existing.ExpirationDate = c.ExpirationDate
	if existing.Status != "active" && existing.Status != "revoked" {
		return nil, nil, fmt.Errorf("cannot RenewAgreement: DataUsageAgreement is in %s state", existing.Status)
	}
	existing.Status = "active"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := RenewedAgreement{
		AggregateID: existing.ID,
		AgreementId: c.AgreementId,
		ExpirationDate: c.ExpirationDate,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}
